File: src/main/java/ru/Models/SignalParametersModel.java
package ru.Models;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.DoubleSummaryStatistics;
import java.util.List;

public class SignalParametersModel {
    private double amplitude;
    private double dc;
    private int bufferedSamplesPerSemiPeriods;
    private double frequency;
    private double maxSignalValue;
    private int minSamples = 20;
    private double minSignalValue;
    private int periods;
    private double rms;
    private int samplesPerSemiPeriods;
    private double[] signal;
    private int zeroTransitionCounter;

    public void calculateParameters(double[] signal) {
        this.signal = signal;
        calculateMinAndMaxValues();
        calculateParameters();
    }

    private void calculateMinAndMaxValues() {
        double min = Double.MAX_VALUE;
        double max = Double.MIN_VALUE;

        for (int i = 0; i < signal.length; i++) {
            if (signal[i] > max) {
                max = signal[i];
            }

            if (signal[i] < min) {
                min = signal[i];
            }
        }

        minSignalValue = min;
        maxSignalValue = max;

        if (frequency > 2) {
            calculateAverageMinAndMaxValues();
        }
    }

    private void calculateAverageMinAndMaxValues() {
        double[] channelData = new double[signal.length];
        List<Double> maxs = new ArrayList<>();
        List<Double> mins = new ArrayList<>();
        int pieces;

        if (frequency < 5) {
            pieces = 2;
        } else if (frequency > 5 && frequency < 10) {
            pieces = 4;
        } else if (frequency > 10 && frequency < 20) {
            pieces = 5;
        } else {
            pieces = 10;
        }

        for (int i = 0, j = 0; i < signal.length; i++) {
            channelData[j++] = signal[i];
        }

        for (int pieceIndex = 0; pieceIndex < pieces; pieceIndex++) {
            double[] pieceOfDate = new double[channelData.length / pieces];
            System.arraycopy(channelData, pieceIndex * pieceOfDate.length, pieceOfDate, 0, pieceOfDate.length);
            DoubleSummaryStatistics statistics = Arrays.stream(pieceOfDate).summaryStatistics();
            maxs.add(statistics.getMax());
            mins.add(statistics.getMin());
        }

        double max = Double.MIN_VALUE;
        double min = Double.MAX_VALUE;
        int indexOfMax = 0;
        int indexOfMin = 0;
        for (int i = 0; i < maxs.size(); i++) {
            if (max < maxs.get(i)) {
                max = maxs.get(i);
                indexOfMax = i;
            }
            if (min > mins.get(i)) {
                min = mins.get(i);
                indexOfMin = i;
            }
        }

        maxs.remove(indexOfMax);
        mins.remove(indexOfMin);
        max = min = 0;

        for (int i = 0; i < maxs.size(); i++) {
            max += maxs.get(i) / maxs.size();
            min += mins.get(i) / mins.size();
        }

        minSignalValue = min;
        maxSignalValue = max;
    }

    private void calculateParameters() {
        amplitude = calculateAmplitude();
        dc = calculateDC();
        rms = calculateRms();
        frequency = calculateFrequency();
    }

    private double calculateAmplitude() {
        return (maxSignalValue - minSignalValue) / 2;
    }

    private double calculateDC() {
        return (maxSignalValue + minSignalValue) / 2;
    }

    private double calculateRms() {
        double summ = 0;
        for (int i = 0; i < signal.length; i++) {
            summ += (signal[i] - dc) * (signal[i] - dc);
        }
        return Math.sqrt(summ / signal.length);
    }

    private double calculateFrequency() {
        double estimatedFrequency = estimateFrequency();
        double accuracyCoefficient = 500;

        if (estimatedFrequency < accuracyCoefficient) {
            return defineFrequency();
        } else {
            return estimatedFrequency;
        }

    }

    private double estimateFrequency() {
        boolean positivePartOfSignal = false;
        double frequency = 0;
        double filteringCoefficient = 1.05;

        for (int i = 0; i < signal.length; i++) {
            if (amplitude + dc >= 0) {
                if (dc >= 0) {
                    if (signal[i] >= dc * filteringCoefficient && !positivePartOfSignal) {
                        frequency++;
                        positivePartOfSignal = true;
                    } else if (signal[i] < dc / filteringCoefficient && positivePartOfSignal) {
                        positivePartOfSignal = false;
                    }
                } else {
                    if (signal[i] >= dc / filteringCoefficient && !positivePartOfSignal) {
                        frequency++;
                        positivePartOfSignal = true;
                    } else if (signal[i] < dc * filteringCoefficient && positivePartOfSignal) {
                        positivePartOfSignal = false;
                    }
                }
            } else if (amplitude + dc < 0 && dc < 0) {
                if (signal[i] < dc * filteringCoefficient && !positivePartOfSignal) {
                    frequency++;
                    positivePartOfSignal = true;
                } else if (signal[i] >= dc / filteringCoefficient && positivePartOfSignal) {
                    positivePartOfSignal = false;
                }
            }
        }

        return frequency;
    }

    private double defineFrequency() {
        int shift = 1_000;
        double firstValue = signal[0] + shift;
        boolean firstPeriod = true;
        boolean positivePartOfSignal = !(firstValue > (dc + shift));
        bufferedSamplesPerSemiPeriods = periods = samplesPerSemiPeriods = zeroTransitionCounter = 0;

        for (int index = 0; index < signal.length; index++) {
            double value = signal[index] + shift;
            double centerOfSignal = dc + shift;

            countSamples();

            if (firstValue >= centerOfSignal) {
                if (value >= centerOfSignal && firstPeriod && (index >= minSamples)) {
                    positivePartOfSignal = true;
                } else if ((value < centerOfSignal && positivePartOfSignal)) {
                    countPeriods();
                    positivePartOfSignal = false;
                    firstPeriod = false;
                } else if (value >= centerOfSignal && !firstPeriod && !positivePartOfSignal && samplesPerSemiPeriods > minSamples) {
                    countPeriods();
                    positivePartOfSignal = true;
                }
            }

            if (firstValue < centerOfSignal) {
                if (value < centerOfSignal && firstPeriod && (index >= minSamples)) {
                    positivePartOfSignal = false;
                } else if ((value > centerOfSignal && !positivePartOfSignal)) {
                    countPeriods();
                    positivePartOfSignal = true;
                    firstPeriod = false;
                } else if (value < centerOfSignal && !firstPeriod && positivePartOfSignal && samplesPerSemiPeriods > minSamples) {
                    countPeriods();
                    positivePartOfSignal = false;
                }
            }
        }

        double samplesPerPeriod = bufferedSamplesPerSemiPeriods == 0 ? 0 : bufferedSamplesPerSemiPeriods / periods;
        double signalFrequency = (samplesPerPeriod == 0 ? 0 : ((double) signal.length / samplesPerPeriod));

        return signalFrequency;
    }

    private void countSamples() {
        if (zeroTransitionCounter >= 1) {
            samplesPerSemiPeriods++;
        }
    }

    private void countPeriods() {
        zeroTransitionCounter++;
        if (zeroTransitionCounter % 2 != 0 && zeroTransitionCounter > 2) {
            bufferedSamplesPerSemiPeriods = samplesPerSemiPeriods;
            periods++;
        }
    }

    public double getAmplitude() {
        return amplitude;
    }

    public double getDc() {
        return dc;
    }

    public double getFrequency() {
        return frequency;
    }

    public double getRms() {
        return rms;
    }
}

File: src/main/java/ru/models/SignalParametersModel.kt
package ru.models

import java.util.*
import kotlin.collections.ArrayList
import kotlin.math.sqrt

class SignalParametersModel {
    var amplitude = 0.0
    var dc = 0.0
    var bufferedSamplesPerSemiPeriods = 0
    var frequency = 0.0
    var maxSignalValue = 0.0
    var minSignalValue = 0.0
    var periods = 0
    var rms = 0.0
    var samplesPerSemiPeriod = 0
    var signal = DoubleArray(0)
    var zeroTransitionCounter = 0

    fun calculateParameters(signal: DoubleArray) {
        this.signal = signal
        calculateMinAndMaxValues()
        amplitude = calculateAmplitude()
        dc = calculateDc()
        rms = calculateRms()
        frequency = calculateFrequency()
    }

    private fun calculateMinAndMaxValues() {
        var max = Double.MIN_VALUE
        var min = Double.MAX_VALUE

        for (value in signal) {
            if (value > max) max = value
            if (value < min) min = value
        }

        maxSignalValue = max
        minSignalValue = min

        if (frequency > 2) calculateAverageMaxAndMinValues()
    }

    private fun calculateAverageMaxAndMinValues() {
        val channelData = DoubleArray(signal.size)
        val maxValues = ArrayList<Double>()
        val minValues = ArrayList<Double>()

        // Define pieces of date
        val pieces = if (frequency <= 5) 2
        else if (frequency > 5 && frequency <= 10) 4
        else if (frequency > 10 && frequency <= 20) 5
        else 10

        // Copy piece of date
        var i = 0
        for (index in 0..signal.size) channelData[i++] = signal[index]

        // Define max and min values
        for (index in 0..pieces) {
            val pieceOfDate = DoubleArray(channelData.size / pieces)
            System.arraycopy(channelData, index * pieceOfDate.size, pieceOfDate, 0, pieceOfDate.size)
            val statistic = Arrays.stream(pieceOfDate).summaryStatistics()
            maxValues.add(statistic.max)
            minValues.add(statistic.min)
        }

        // Filter noise
        var max = Double.MIN_VALUE
        var min = Double.MAX_VALUE
        var indexOfMax = 0
        var indexOfMin = 0
        for (index in 0..maxValues.size) {
            if (max < maxValues[index]) {
                max = maxValues[index]
                indexOfMax = index
            }
            if (min > minValues[index]) {
                min = minValues[index]
                indexOfMin = index
            }
        }
        maxValues.removeAt(indexOfMax)
        minValues.removeAt(indexOfMin)

        // Calculate average of maxValues and minValues
        max = 0.0
        min = 0.0
        for (index in 0..maxValues.size) {
            max += maxValues[index] / maxValues.size
            min += minValues[index] / minValues.size
        }

        maxSignalValue = max
        minSignalValue = min
    }

    private fun calculateAmplitude(): Double {
        return (maxSignalValue - minSignalValue) / 2
    }

    private fun calculateDc(): Double {
        return (maxSignalValue + minSignalValue) / 2
    }

    private fun calculateRms(): Double {
        var sum = 0.0
        for (value in signal) sum += (value - dc) * (value - dc)
        return sqrt(sum / signal.size)
    }

    private fun calculateFrequency(): Double {
        val estimatedFrequency = estimateFrequency()
        val accuracyCoefficient = 500
        if (estimatedFrequency < accuracyCoefficient) return defineFrequency()
        else return estimatedFrequency.toDouble()
    }

    private fun estimateFrequency(): Int {
        var isPositivePartOfSignal = false
        var frequency = 0
        val filteringCoefficient = 1.05
        val lowerBound = if (dc < 0) dc * filteringCoefficient else dc / filteringCoefficient
        val upperBound = if (dc < 0) dc / filteringCoefficient else dc * filteringCoefficient

        for (value in signal) {
            if (value >= upperBound && !isPositivePartOfSignal) {
                frequency++
                isPositivePartOfSignal = true
            } else if (value < lowerBound && isPositivePartOfSignal) {
                isPositivePartOfSignal = false
            }
        }

        return frequency
    }

    private fun defineFrequency(): Double {
        val shift = 1000
        val isFirstPeriod = false
        val isPositivePartOfSignal = true
        val filteringCoefficient = 1.05
        bufferedSamplesPerSemiPeriods = 0
        periods = 0
        samplesPerSemiPeriod = 0
        zeroTransitionCounter = 0

        for (value in signal) {
            val shiftedValue = value + shift
            val centerOfSignal = dc + shift
            val lowerBound = if (centerOfSignal < 0) centerOfSignal * filteringCoefficient else centerOfSignal / filteringCoefficient
            val upperBound = if (centerOfSignal < 0) centerOfSignal / filteringCoefficient else centerOfSignal * filteringCoefficient

            if (zeroTransitionCounter >= 1) samplesPerSemiPeriod++

            if (shiftedValue >= upperBound && isFirstPeriod) {
            } else if (shiftedValue < lowerBound && isPositivePartOfSignal) {
                countPeriods()
            } else if (value >= upperBound && !isFirstPeriod && !isPositivePartOfSignal) {
                countPeriods()
            }
        }

        val samplesPerPeriod = if (bufferedSamplesPerSemiPeriods == 0) 0 else bufferedSamplesPerSemiPeriods / periods
        return if (samplesPerPeriod == 0) 0.0 else signal.size.toDouble() / samplesPerPeriod.toDouble()
    }

    private fun countPeriods() {
        zeroTransitionCounter++
        if (zeroTransitionCounter % 2 !=0 && zeroTransitionCounter > 2) {
            bufferedSamplesPerSemiPeriods = samplesPerSemiPeriod
            periods++
        }
    }
}